File: DataStructures/src/me/divium/LinkedList.java
package me.divium;

/*
    + Создание списка
   +- Удаление элемента (Не работает удаление одинаквоых элементов, когда они стоят рядом)
    + Добавление/вставка элемента
    + Поиск по ключу
    + Вывод списка на экран
 */

public class LinkedList<T> {

    private Node node;
    private int size;

    private class Node {
        T data;
        Node next;

        public Node() {
            this.data = null;
            this.next = null;
        }

        public Node(T data) {
            this.data = data;
            this.next = null;
        }

        @Override
        public String toString() {
            return data.toString();
        }
    }

    public LinkedList() {
        this.node = new Node();
        this.size = 1;
    }

    /**
     * Всталяет объект в конец списка
     * @param data Объект
     */
    public void add(T data) {
        var currentNode = this.node;

        //Первичное добавление элемента в список
        if (currentNode.data == null) {
            currentNode.data = data;
            return;
        }

        //Добавление элемента в конец списка
        while (currentNode.next != null)
            currentNode = currentNode.next;

        currentNode.next = new Node(data);
        increaseSize();
    }

    /**
     * Вставляет объект на указываемый индекс. Нельзя вставить в конец списка
     * @param data Вставляемый объект
     * @param index Индекс
     */
    public void insert(T data, int index) {
        validateIndex(index);

        //Вставка в если в списке ничего нет
        if (this.node.data == null) {
            add(data);
            return;
        }

        //Вставка в начало списка
        if (index == 0) {
            Node movedNode = this.node;
            this.node = new Node(data);
            this.node.next = movedNode;
            return;
        }

        Node currentNode = this.node;
        for (int i = 0; i + 1 != index; i++)
            currentNode = currentNode.next;

        Node nextNode = currentNode.next;
        currentNode.next = new Node(data);
        currentNode.next.next = nextNode;
        increaseSize();
    }

    public void remove(int index) {
        removeByIndex(index);
    }

    public void remove(T data) {
        removeByObject(data);
    }

    public void removeAll(T data) {
        while (this.contains(data))
            remove(data);
    }

    public boolean contains(T data) {
        var currentNode = this.node;

        while (currentNode != null) {
            if (currentNode.data.equals(data))
                return true;
            currentNode = currentNode.next;
        }

        return false;
    }

    public int size() {
        return size;
    }

    @Override
    public String toString() {
        var builder = new StringBuilder();
        var currentNode = this.node;

        while (currentNode != null) {
            builder.append(currentNode.data.toString()).append(" -> ");
            currentNode = currentNode.next;
        }

        builder.append("null");
        return builder.toString();
    }

//===============================================================================

    private void increaseSize() {
        this.size++;
    }

    private void decreaseSize() {
        this.size--;
    }

    private void validateIndex(int index) {
        if (index < 0 || index > size - 1)
            throw new IllegalArgumentException("Index out of bounds");
    }

    private void removeFirstNode() {
        this.node = this.node.next;
        decreaseSize();
    }

    private void removeNextNode(Node currentNode) {
        var deleteNode = currentNode.next;
        currentNode.next = deleteNode.next;
        decreaseSize();
    }

    private void removeByIndex(int index) {
        validateIndex(index);

        if (index == 0) {
            removeFirstNode();
            return;
        }

        var currentNode = this.node;
        int currentIndex = 0;

        while (currentIndex + 1 != index) {
            currentNode = currentNode.next;
            currentIndex++;
        }

        removeNextNode(currentNode);
    }

    private void removeByObject(T data) {
        if (this.node.data.equals(data)) {
            removeFirstNode();
            return;
        }

        var currentNode = this.node;
        while (currentNode.next != null) {
            if (currentNode.next.data.equals(data)) {
                removeNextNode(currentNode);
                return;
            }
            currentNode = currentNode.next;
        }
    }

}

File: DataStructersCpp/LinkedList.cpp
﻿#pragma once

#include "LinkedList.h"

using namespace std;

#pragma region Node constructors and destructors
template<class T> LinkedList<T>::Node::Node(T data) {
	this->data = data;
	this->next = nullptr;
}

template<class T> LinkedList<T>::Node::~Node() {
	if (next != nullptr)
		delete next;
}
#pragma endregion

#pragma region Linked List Private Methods
template<class T> void LinkedList<T>::increaseSize() {
	listSize++;
}

template<class T> void LinkedList<T>::decreaseSize() {
	listSize--;
}

template<class T> void const LinkedList<T>::validateIndex(const int index) {
	if (index < 0 || index > listSize)
		throw new out_of_range("Index out of bounds");
}

template<class T> void LinkedList<T>::deleteFirstNode() {
	//сохраняем указатель на 2-ую ноду
	auto secondNode = this->node->next;

	//удаляем первую ноду, присваиваем нулевой указатель чтобы не запустилась цепная реакция
	this->node->next = nullptr;
	delete this->node;

	//делаем вторую ноду первой
	this->node = secondNode;
	decreaseSize();
}

template<class T> void LinkedList<T>::deleteNextNode(Node* currentNode) {
	//пропускаем ноду которую надо удалить
	auto deleteNode = currentNode->next;
	currentNode->next = deleteNode->next;

	//теперь удаляем её
	deleteNode->next = nullptr;
	delete deleteNode;
	decreaseSize();
}
#pragma endregion

#pragma region Linked List Constructors and destructors
template<class T> LinkedList<T>::LinkedList() {
	this->node = nullptr;
	this->listSize = 0;
}

template<class T> LinkedList<T>::~LinkedList() {
	delete node;
}
#pragma endregion

#pragma region Linked List Public Methods
template<class T> void LinkedList<T>::add(T data) {
	//Первичное добавление элемента в список
	if (this->node == nullptr) {
		this->node = new Node(data);
		return;
	}

	Node* currentNode = this->node;

	//Добавление элемента в конец списка
	//Поиск последней ноды
	while (currentNode->next != nullptr)
		currentNode = currentNode->next;

	currentNode->next = new Node(data);
	increaseSize();
}

template<class T> void LinkedList<T>::insert(T data, const int index) {
	validateIndex(index);

	if (index == 0) {
		Node* movedNode = this->node;
		this->node = new Node(data);
		this->node->next = movedNode;
		return;
	}

	int currentIndex = 0;
	Node* currentNode = this->node;

	while (currentIndex + 1 != index) {
		currentNode = currentNode->next;
		currentIndex++;
	}

	Node* nextNode = currentNode->next;
	currentNode->next = new Node(data);
	currentNode->next->next = nextNode;
	increaseSize();
}

template<class T> void LinkedList<T>::insert(T data, T after) {
	Node* currentNode = this->node;

	while (currentNode != nullptr && currentNode->data.equals(after))
		currentNode = currentNode->next;

	Node* nextNode = currentNode->next;
	currentNode->next = new Node(data);
	currentNode->next->next = nextNode;
	increaseSize();
}

template<class T> bool const LinkedList<T>::contains(T data) {
	Node* currentNode = this->node;

	while (currentNode != nullptr) {
		if (currentNode->data.equals(data))
			return true;

		currentNode = currentNode->next;
	}

	return false;
}

template<class T> bool const LinkedList<T>::contains(T data, bool (*lambda)(T& object1, T& object2)) {
	Node* currentNode = this->node;

	while (currentNode != nullptr) {
		if (lambda(currentNode->data, data))
			return true;

		currentNode = currentNode->next;
	}

	return false;
}

template<class T> void LinkedList<T>::remove(const int index) {
	validateIndex(index);

	if (index == 0) {
		deleteFirstNode();
		return;
	}

	auto currentNode = this->node;
	for (int i = 0; i + 1 != index; i++)
		currentNode = currentNode->next;

	deleteNextNode(currentNode);
}

template<class T> void LinkedList<T>::remove(T data) {
	if (this->node == nullptr)
		return;

	if (this->node->data.equals(data)) {
		deleteFirstNode();
		return;
	}

	auto currentNode = this->node;
	while (currentNode->next != nullptr) {
		if (currentNode->next->data.equals(data)) {
			deleteNextNode(currentNode);
			return;
		}

		currentNode = currentNode->next;
	}

}

template<class T> void LinkedList<T>::remove(T data, bool (*lambda)(T& object1, T& object2)) {
	if (this->node == nullptr)
		return;

	if (lambda(this->node->data, data)) {
		deleteFirstNode();
		return;
	}

	auto currentNode = this->node;
	while (currentNode->next != nullptr) {
		if (lambda(currentNode->next->data, data)) {
			deleteNextNode(currentNode);
			return;
		}
		deleteNextNode(currentNode);
	}
}

template<class T> void LinkedList<T>::removeAll(T data) {
	while (this->contains(data))
		remove(data);
}

template<class T> void LinkedList<T>::removeAll(T data, bool (*lambda)(T& object1, T& object2)) {
	while (this->contains(data, lambda))
		remove(data, lambda);
}

template<class T> size_t const LinkedList<T>::size() {
	return this->listSize;
}

template<class T> string const LinkedList<T>::toString() {
	stringstream stream;
	Node* currentNode = this->node;

	while (currentNode != nullptr) {
		stream << currentNode->data.toString() << "->";
		currentNode = currentNode->next;
	}

	stream << "null";
	return stream.str();
}

template<class T> string const LinkedList<T>::toString(string(*lambda)(T& object)) {
	stringstream stream;
	Node* currentNode = this->node;

	while (currentNode != nullptr) {
		stream << lambda(currentNode->data) << "->";
		currentNode = currentNode->next;
	}

	stream << "null";
	return stream.str();
}
#pragma endregion